### PR TITLE
Provide access to F# integration tests

### PR DESCRIPTION
**Pull request by psfinaki**
We are ramping integration testing for F# and need internal access to some APIs, similar to how TypeScript or Razor have them.

The desired test in question is [here](https://github.com/dotnet/fsharp/pull/14764/files#diff-cb15dc9adef6f54d7312918096dbb1da9291da38a3b82d23ae12a1e22ef4095b). This currently fails [here](https://github.com/dotnet/fsharp/pull/14764/files#diff-c299ff93d2440ee88797821ae6c712f40b698914c9bf704baef2b4f5ff4868f4) when trying to access workspaces stuff:
image

We are really just trying to do what Roslyn repo does itself for C# [here](https://github.com/dotnet/roslyn/blob/af816d93f022d3f6d76284ea87ebc3fa98cf4705/src/VisualStudio/IntegrationTest/New.IntegrationTests/CSharp/CSharpNavigateTo.cs).

### CODE REVIEW

**Comment by Cosifne** [comment]
I think the general strategy to share code is moving things to the external access pattern instead of directly using IVT. 
Is that difficult to be done?

**Comment by davidwengier** [comment]
> similar to how TypeScript or Razor have them

We (Razor) don't have IVT in this manner, but use an external access layer as @Cosifne mentioned, though its worth noting we don't actually use that for our integration tests. So for example for Go To Definition, rather than trying to reach into the internals of Roslyn to wait for an async operation to finish, we just [wait for the active window](https://github.com/dotnet/razor/blob/main/src/Razor/test/Microsoft.VisualStudio.Razor.IntegrationTests/GoToDefinitionTests.cs#L40) to be the one we expect. That helper method [just polls](https://github.com/dotnet/razor/blob/main/src/Razor/test/Microsoft.VisualStudio.Razor.IntegrationTests/InProcess/EditorInProcess_Text.cs#L122).

File: src/EditorFeatures/Core/Microsoft.CodeAnalysis.EditorFeatures.csproj [hunk 1]
```diff
@@ -94,6 +94,7 @@
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.AspNetCore" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.FSharp" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.FSharp.UnitTests" />
+    <InternalsVisibleTo Include="FSharp.Editor.IntegrationTests" Key="$(FSharpKey)" />
```
Comment (sharwell) [inline src/EditorFeatures/Core/Microsoft.CodeAnalysis.EditorFeatures.csproj:97]: ❗ This must run through the external access layer. You can create a new project for F# integration tests, or you can reuse the existing F# external access project.

File: src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj [hunk 2]
```diff
@@ -93,6 +93,7 @@
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.Test.Utilities2" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.FSharp" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.FSharp.UnitTests" />
+    <InternalsVisibleTo Include="FSharp.Editor.IntegrationTests" Key="$(FSharpKey)" />
```
Comment (sharwell) [inline src/Features/Core/Portable/Microsoft.CodeAnalysis.Features.csproj:96]: ❗ This must run through the external access layer. You can create a new project for F# integration tests, or you can reuse the existing F# external access project.

File: src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj [hunk 3]
```diff
@@ -132,6 +132,7 @@
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.Test.Utilities2" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.FSharp" />
     <InternalsVisibleTo Include="Microsoft.CodeAnalysis.ExternalAccess.FSharp.UnitTests" />
+    <InternalsVisibleTo Include="FSharp.Editor.IntegrationTests" Key="$(FSharpKey)" />
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/Microsoft.CodeAnalysis.Workspaces.csproj:135]: ❗ This must run through the external access layer. You can create a new project for F# integration tests, or you can reuse the existing F# external access project.

**Review by sharwell — CHANGES_REQUESTED** [review]

**Comment by sharwell** [comment]
> For example, when testing Go To Definition rather than trying to reach into the internals of Roslyn to wait for an async operation to finish, we just [wait for the active window](https://github.com/dotnet/razor/blob/main/src/Razor/test/Microsoft.VisualStudio.Razor.IntegrationTests/GoToDefinitionTests.cs#L40) to be the one we expect. That helper method [just polls](https://github.com/dotnet/razor/blob/main/src/Razor/test/Microsoft.VisualStudio.Razor.IntegrationTests/InProcess/EditorInProcess_Text.cs#L122). We don't currently have any NavigateTo tests (but we probably should!) but on the face of it, I see no reason the test you linked to couldn't be written the same way as our GTD test.

These strategies have totally different performance characteristics. When you wait for Roslyn asynchronous operations to complete, it immediately forces all delayed operations run to completion. See [`IExpeditableDelaySource`](https://github.com/dotnet/roslyn/blob/main/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/TestHooks/IExpeditableDelaySource.cs).

**Comment by davidwengier** [comment]
> somewhere between a code smell and an outright design bug

I would make the same statement about the fact that any of the Razor integration tests need to exist in the first place. When it comes to them, I'm of the opinion that if it's crazy and it works, then it's not crazy 😄

**Comment by psfinaki** [comment]
Hah okay actually looks like we already have all the dependendies we are interested in, it's just that the integration tests project is new and doesn't have access to the external access layer yet - so fixing just this.

**Comment by psfinaki** [comment]
So things makes sense now. @sharwell @Cosifne this should be sufficient for now.

**Review by 0101 — APPROVED** [review]

**Review by sharwell — APPROVED** [review]
This is now fine, but we should squash-merge this to avoid churn.

**Pull request merged**